From 26e8d235451ce07f080e5c2f579ad428b922fc9a Mon Sep 17 00:00:00 2001
From: "Jesse A. Hernandez" <jahphys@gmail.com>
Date: Thu, 24 Sep 2026 20:20:09 -0400
Subject: [PATCH 1/5] Hist2D: add CosTheta histTypes

Adds CosThetaVsEgamma, CosThetaVst and CosThetaVsMass (cos theta of the particle-list system in the recoil rest frame). t for MassVst becomes |(beam-recoil)^2|.

Upstream: JeffersonLab/halld_sim@bcff7a5c4e8453b1e75e83facdc6f2ad95ba5719 (4.54.0); upstream copyright retained.
Source: the author's JLab ifarm MC area (halld_sim-2019_11-ver01_13; the four run-period copies are identical).
---
 src/libraries/AMPTOOLS_AMPS/Hist2D.cc | 73 ++++++++++++++++++++++++---
 1 file changed, 65 insertions(+), 8 deletions(-)

diff --git a/src/libraries/AMPTOOLS_AMPS/Hist2D.cc b/src/libraries/AMPTOOLS_AMPS/Hist2D.cc
index 9a0ba8d..fa4841b 100644
--- a/src/libraries/AMPTOOLS_AMPS/Hist2D.cc
+++ b/src/libraries/AMPTOOLS_AMPS/Hist2D.cc
@@ -5,6 +5,9 @@
 #include <sstream>
 #include <cstdlib>
 
+#include "TLorentzVector.h"
+#include "TLorentzRotation.h"
+
 #include "IUAmpTools/Kinematics.h"
 #include "AMPTOOLS_AMPS/Hist2D.h"
 
@@ -39,7 +42,7 @@ UserAmplitude< Hist2D >( args )
 		exit(1);
 	}		
 
-	if(histType != "MassVsEgamma" && histType != "MassVst") {
+	if(histType != "MassVsEgamma" && histType != "MassVst" && histType != "CosThetaVsEgamma" && histType != "CosThetaVst" && histType != "CosThetaVsMass") {
 		cout<<"Type of 2D histogram is not currently supported, please add necessary kinematics and options to Hist2D amplitude"<<endl;
 		exit(1);
 	}
@@ -55,29 +58,83 @@ Hist2D::calcAmplitude( GDouble** pKin ) const {
   
 	TLorentzVector target  ( 0., 0., 0., 0.938);
 	TLorentzVector beam   ( pKin[0][1], pKin[0][2], pKin[0][3], pKin[0][0] ); 
-	TLorentzVector recoil ( pKin[1][1], pKin[1][2], pKin[1][3], pKin[1][0] ); 
-	
+	TLorentzVector recoil ( pKin[1][1], pKin[1][2], pKin[1][3], pKin[1][0] );        
+	TLorentzVector p1 ( pKin[2][1], pKin[2][2], pKin[2][3], pKin[2][0] ); 
+	TLorentzVector p2 ( pKin[3][1], pKin[3][2], pKin[3][3], pKin[3][0] ); 
+
 	// compute particle P4 sum for invariant mass
-	TLorentzVector sum;
+	TLorentzVector res;
 	for(uint i=0; i<particleList.length(); i++) {
 		string num; num += particleList[i];
 		int index = atoi(num.c_str());
 		TLorentzVector particleP4 ( pKin[index][1], pKin[index][2], pKin[index][3], pKin[index][0] ); 
-		sum += particleP4;
+		res += particleP4;
 	}
 	
 	double beamE = beam.E();
-	double t = fabs((sum - beam).M2()); 
+	double t = fabs((beam - recoil).M2()); 
 
 	double userVarX = 0;
-	double userVarY = sum.M();
+	double userVarY = 0;
 	if(histType == "MassVsEgamma") {
-		userVarX = beamE;
+		userVarY = res.M();
+	        userVarX = beamE;
 	}
 	if(histType == "MassVst") {
+                userVarY = res.M();
 		userVarX = t;
 	}
 
+	// //COM frame
+	// TLorentzVector com = beam + target;
+	// TLorentzRotation comRestBoost ( -com.BoostVector() );
+
+	// TLorentzVector beam_com   = comRestBoost * beam;
+	// TLorentzVector recoil_com = comRestBoost * recoil;
+	// TLorentzVector res_com = comRestBoost * res;
+	// TLorentzVector p1_com = comRestBoost * p1;
+	// TLorentzVector p2_com = comRestBoost * p2;
+	
+	//Resonance Rest Frame from COM
+	TLorentzRotation resRestBoost( -res.BoostVector() );
+	TLorentzVector beam_res   = resRestBoost * beam;
+	TLorentzVector recoil_res = resRestBoost * recoil;
+	TLorentzVector p1_res = resRestBoost * p1;
+	TLorentzVector p2_res = resRestBoost * p2;	
+	
+	// helicity frame: z-axis is propagation of resonance X => opposite recoil proton in X rest frame
+	TVector3 z = 1 * res.Vect().Unit();
+
+	// y axis perpendicular to production plane
+	TVector3 y = (beam.Vect().Cross(res.Vect())).Unit();
+
+	TVector3 x = y.Cross(z);
+  
+	TVector3 p2_p ( p2_res.Vect().Dot(x),
+			 p2_res.Vect().Dot(y),
+			 p2_res.Vect().Dot(z) );
+  
+	double cosTheta = p2_p.Pz() / p2_p.Mag();
+
+	//cout << "Beam P(initial, com): " << beam.P() << ", " << beam_com.P() << endl;
+	//cout << "Checking for CoM frame(Recoil, Res): (" << recoil_com.Pz() << ", " << res_com.Pz() << ")" << endl;
+       
+	//double phi = angles.Phi();
+	if(histType == "CosThetaVsEgamma") {
+	        userVarY = cosTheta;
+		userVarX = beamE;
+	}
+
+	if(histType == "CosThetaVst") {
+	        userVarY = cosTheta;
+		userVarX = t;
+	}
+
+	if(histType == "CosThetaVsMass") {
+	        userVarY = cosTheta;
+		userVarX = res.M();
+	}	
+
 	// weighted model of intensity from histogram 
 	GDouble W = 0.; // initialized to zero
 
-- 
2.50.1 (Apple Git-155)

